src/layout/umap.c

```diff
@@ -0,0 +1,1038 @@
+#include "igraph_interface.h"
+#include "igraph_lapack.h"
+#include "igraph_matrix.h"
+#include "igraph_random.h"
+#include "igraph_nongraph.h"
+
+
+/* rho is just the size of the distance from each vertex and its closest neighbor */
+/* sigma is the the decay from each vertex, depends on its rho and the rest of its neighbor
+ * distances */
+
+/* Find sigma for this vertex by binary search */
+static igraph_error_t igraph_i_umap_find_sigma(const igraph_vector_t *distances,
+        const igraph_vector_int_t *eids,
+        igraph_real_t rho, igraph_real_t *sigma_p, igraph_real_t target) {
+
+    igraph_real_t sigma = 1;
+    igraph_real_t sum;
+    igraph_real_t tol = 0.01;
+    igraph_integer_t maxiter = 100;
+    igraph_integer_t no_of_neis = igraph_vector_int_size(eids);
+    igraph_integer_t eid;
+    igraph_real_t step = sigma;
+    igraph_integer_t seen_max = 0;
+
+    /* Binary search */
+    for (igraph_integer_t iter = 0; iter < maxiter; iter++) {
+        sum = 0;
+        for (igraph_integer_t j = 0; j < no_of_neis; j++) {
+            eid = VECTOR(*eids)[j];
+            sum += exp(-(VECTOR(*distances)[eid] - rho) / sigma);
+        }
+
+#ifdef UMAP_DEBUG
+        printf("SIGMA function (no_of_neis = %" IGRAPH_PRId ")- sum: %g, "
+               "target: %g, rho: %g, sigma: %g\n", no_of_neis, sum, target, rho, sigma);
+#endif
+
+        if (sum < target) {
+            /* going back up after having seen an upper bound */
+            if (seen_max == 1) {
+                step /= 2;
+            /* we need to go up but have not seen an upper bound yet
+             * first iteration we want to increase by sigma, else we must come from
+             * below, so we are sitting at 2 * step, we want to move to 4 * step */
+            } else if (iter > 0) {
+                step *= 2;
+            }
+            sigma += step;
+        /* overshooting, we have definitely seen the max */
+        } else {
+            seen_max = 1;
+            step /= 2;
+            sigma -= step;
+        }
+
+        /* Check for convergence */
+        if (fabs(sum - target) < tol) {
+            break;
+        }
+    }
+
+    *sigma_p = sigma;
+
+    return IGRAPH_SUCCESS;
+}
+
+
+/* Convert the graph with distances into a probability graph with exponential decay */
+/* NOTE: this is funny because the distance was a correlation to start with... ?? */
+static igraph_error_t igraph_i_umap_find_prob_graph(const igraph_t *graph,
+        const igraph_vector_t *distances, igraph_vector_t *umap_weights) {
+
+    igraph_integer_t no_of_nodes = igraph_vcount(graph);
+    igraph_integer_t no_of_edges = igraph_ecount(graph);
+    igraph_integer_t no_of_neis, eid;
+    igraph_vector_int_t eids, weight_seen;
+    igraph_real_t rho, dist_max, dist, sigma, weight, weight_inv, sigma_target;
+
+    /* if the original graph is unweighted, probabilities are 1 throughout */
+    if (distances == NULL) {
+        no_of_neis = igraph_vector_size(umap_weights);
```

Comment:
I believe this line is not needed (variable is never used).

Reply:
good catch, I deleted it now